failing TCP tests

Some of our TCP transport tests are failing randomly.

Failing run: https://github.com/libp2p/go-libp2p/runs/6488638370?check_suite_focus=true

This is probably easiest to debug on the branch on which we merge the transport test suite into go-libp2p (the failure also occurred before, now we just have all the code in one place): https://github.com/libp2p/go-libp2p/pull/1496

```
=== RUN   TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress50Conn10Stream50Msg
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:60: stream reset
      stream_suite.go:60: stream reset
      stream_suite.go:60: stream reset
      stream_suite.go:60: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:143: s.Write(buf): read tcp4 127.0.0.1:50539->127.0.0.1:50538: read: connection reset by peer
      stream_suite.go:143: s.Write(buf): read tcp4 127.0.0.1:50539->127.0.0.1:50538: read: connection reset by peer
      stream_suite.go:143: s.Write(buf): read tcp4 127.0.0.1:50539->127.0.0.1:50538: read: connection reset by peer
      stream_suite.go:143: s.Write(buf): read tcp4 127.0.0.1:50539->127.0.0.1:50538: read: connection reset by peer
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:143: s.Write(buf): stream reset
      stream_suite.go:143: s.Write(buf): stream reset
      stream_suite.go:143: s.Write(buf): stream reset
      stream_suite.go:143: s.Write(buf): stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
  === RUN   TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress1Conn1000Stream10Msg
  === RUN   TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress1Conn100Stream100Msg10MB
  === RUN   TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStreamOpenStress
  === RUN   TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStreamReset
  === RUN   TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestProtocols#01
  === RUN   TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestBasic#01
  === RUN   TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestCancel#01
  === RUN   TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestPingPong#01
  === RUN   TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress1Conn1Stream1Msg#01
  === RUN   TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress1Conn1Stream100Msg#01
  === RUN   TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress1Conn100Stream100Msg#01
  === RUN   TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress50Conn10Stream50Msg#01
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:143: s.Write(buf): write tcp4 127.0.0.1:50776->127.0.0.1:50774: write: no buffer space available
      stream_suite.go:143: s.Write(buf): write tcp4 127.0.0.1:50776->127.0.0.1:50774: write: no buffer space available
      stream_suite.go:143: s.Write(buf): write tcp4 127.0.0.1:50776->127.0.0.1:50774: write: no buffer space available
      stream_suite.go:54: read tcp4 127.0.0.1:50722->127.0.0.1:50721: read: connection reset by peer
      stream_suite.go:54: read tcp4 127.0.0.1:50722->127.0.0.1:50721: read: connection reset by peer
      stream_suite.go:54: read tcp4 127.0.0.1:50722->127.0.0.1:50721: read: connection reset by peer
      stream_suite.go:54: read tcp4 127.0.0.1:50722->127.0.0.1:50721: read: connection reset by peer
      stream_suite.go:54: read tcp4 127.0.0.1:50722->127.0.0.1:50721: read: connection reset by peer
      stream_suite.go:54: read tcp4 127.0.0.1:50722->127.0.0.1:50721: read: connection reset by peer
      stream_suite.go:54: read tcp4 127.0.0.1:50722->127.0.0.1:50721: read: connection reset by peer
      stream_suite.go:54: read tcp4 127.0.0.1:50722->127.0.0.1:50721: read: connection reset by peer
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:156: io.ReadFull(s, buf2): stream reset
      stream_suite.go:54: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
      stream_suite.go:83: stream reset
  === RUN   TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress1Conn1000Stream10Msg#01
  === RUN   TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress1Conn100Stream100Msg10MB#01
  === RUN   TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStreamOpenStress#01
  === RUN   TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStreamReset#01
  --- FAIL: TestTcpTransport (8.01s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestProtocols (0.00s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestBasic (0.00s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestCancel (0.00s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestPingPong (0.01s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress1Conn1Stream1Msg (0.00s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress1Conn1Stream100Msg (0.00s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress1Conn100Stream100Msg (0.31s)
      --- FAIL: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress50Conn10Stream50Msg (0.65s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress1Conn1000Stream10Msg (0.55s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress1Conn100Stream100Msg10MB (0.27s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStreamOpenStress (2.66s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStreamReset (0.05s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestProtocols#01 (0.00s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestBasic#01 (0.00s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestCancel#01 (0.00s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestPingPong#01 (0.01s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress1Conn1Stream1Msg#01 (0.00s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress1Conn1Stream100Msg#01 (0.00s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress1Conn100Stream100Msg#01 (0.22s)
      --- FAIL: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress50Conn10Stream50Msg#01 (0.58s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress1Conn1000Stream10Msg#01 (0.32s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStress1Conn100Stream100Msg10MB#01 (0.25s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStreamOpenStress#01 (2.06s)
      --- PASS: TestTcpTransport/github.com/libp2p/go-libp2p/p2p/transport/testsuite.SubtestStreamReset#01 (0.05s)

--- a/p2p/transport/testsuite/stream_suite.go
+++ b/p2p/transport/testsuite/stream_suite.go
@@ -7,6 +7,7 @@ import (
 	"io"
 	"io/ioutil"
 	"os"
+	"runtime"
 	"strconv"
 	"sync"
 	"testing"
@@ -417,19 +418,15 @@ func SubtestStress1Conn100Stream100Msg(t *testing.T, ta, tb transport.Transport,
 	})
 }
 
-func SubtestStress50Conn10Stream50Msg(t *testing.T, ta, tb transport.Transport, maddr ma.Multiaddr, peerA peer.ID) {
+func SubtestStressManyConn10Stream50Msg(t *testing.T, ta, tb transport.Transport, maddr ma.Multiaddr, peerA peer.ID) {
+	connNum := 5
+	if runtime.GOOS == "linux" {
+		// Linux can handle a higher number of conns here than other platforms in CI.
+		// See https://github.com/libp2p/go-libp2p/issues/1498.
+		connNum = 50
+	}
 	SubtestStress(t, ta, tb, maddr, peerA, Options{
-		ConnNum:   50,
-		StreamNum: 10,
-		MsgNum:    50,
-		MsgMax:    100,
-		MsgMin:    100,
-	})
-}
-
-func SubtestStress5Conn10Stream50Msg(t *testing.T, ta, tb transport.Transport, maddr ma.Multiaddr, peerA peer.ID) {
-	SubtestStress(t, ta, tb, maddr, peerA, Options{
-		ConnNum:   5,
+		ConnNum:   connNum,
 		StreamNum: 10,
 		MsgNum:    50,
 		MsgMax:    100,

--- a/p2p/transport/testsuite/stream_suite.go
+++ b/p2p/transport/testsuite/stream_suite.go
@@ -427,6 +427,16 @@ func SubtestStress50Conn10Stream50Msg(t *testing.T, ta, tb transport.Transport, 
 	})
 }
 
+func SubtestStress5Conn10Stream50Msg(t *testing.T, ta, tb transport.Transport, maddr ma.Multiaddr, peerA peer.ID) {
+	SubtestStress(t, ta, tb, maddr, peerA, Options{
+		ConnNum:   5,
+		StreamNum: 10,
+		MsgNum:    50,
+		MsgMax:    100,
+		MsgMin:    100,
+	})
+}
+
 func SubtestStress1Conn1000Stream10Msg(t *testing.T, ta, tb transport.Transport, maddr ma.Multiaddr, peerA peer.ID) {
 	SubtestStress(t, ta, tb, maddr, peerA, Options{
 		ConnNum:   1,

--- a/p2p/transport/testsuite/utils_suite.go
+++ b/p2p/transport/testsuite/utils_suite.go
@@ -21,7 +21,7 @@ var Subtests = []func(t *testing.T, ta, tb transport.Transport, maddr ma.Multiad
 	SubtestStress1Conn1Stream1Msg,
 	SubtestStress1Conn1Stream100Msg,
 	SubtestStress1Conn100Stream100Msg,
-	SubtestStress5Conn10Stream50Msg,
+	SubtestStressManyConn10Stream50Msg,
 	SubtestStress1Conn1000Stream10Msg,
 	SubtestStress1Conn100Stream100Msg10MB,
 	SubtestStreamOpenStress,
@@ -36,13 +36,6 @@ func SubtestTransport(t *testing.T, ta, tb transport.Transport, addr string, pee
 	maddr, err := ma.NewMultiaddr(addr)
 	if err != nil {
 		t.Fatal(err)
-	}
-
-	if runtime.GOOS == "linux" {
-		// Only run this test on Linux since macOS runs into buffering issues on CI
-		// with this many connections. See
-		// https://github.com/libp2p/go-libp2p/issues/1498.
-		Subtests = append(Subtests, SubtestStress50Conn10Stream50Msg)
 	}
 
 	for _, f := range Subtests {

--- a/p2p/transport/testsuite/utils_suite.go
+++ b/p2p/transport/testsuite/utils_suite.go
@@ -21,7 +21,7 @@ var Subtests = []func(t *testing.T, ta, tb transport.Transport, maddr ma.Multiad
 	SubtestStress1Conn1Stream1Msg,
 	SubtestStress1Conn1Stream100Msg,
 	SubtestStress1Conn100Stream100Msg,
-	SubtestStress50Conn10Stream50Msg,
+	SubtestStress5Conn10Stream50Msg,
 	SubtestStress1Conn1000Stream10Msg,
 	SubtestStress1Conn100Stream100Msg10MB,
 	SubtestStreamOpenStress,
@@ -37,6 +37,14 @@ func SubtestTransport(t *testing.T, ta, tb transport.Transport, addr string, pee
 	if err != nil {
 		t.Fatal(err)
 	}
+
+	if runtime.GOOS == "linux" {
+		// Only run this test on Linux since macOS runs into buffering issues on CI
+		// with this many connections. See
+		// https://github.com/libp2p/go-libp2p/issues/1498.
+		Subtests = append(Subtests, SubtestStress50Conn10Stream50Msg)
+	}
+
 	for _, f := range Subtests {
 		t.Run(getFunctionName(f), func(t *testing.T) {
 			f(t, ta, tb, maddr, peerA)

